.github/workflows/main.yml

```diff
@@ -13,24 +13,26 @@ jobs:
         id: set-matrix
         shell: bash
         run: |
+          define_matrix () {
+            comm -12 "$2" "$3" >"$4" &&
+            cat "$4" &&
+            sed -e 's/\/$//' -e 's/.*/"&", /' -e '$s/, $//' <"$4" >list.txt &&
+            echo "::set-output name=$1::[$(test ! -s list.txt && echo '"empty"' || tr -d '\n' <list.txt)]"
+          }
+
           git ls-files \*/PKGBUILD | sed 's|[^/]*$||' | sort >directories.txt &&
           git diff ${{github.event.pull_request.base.sha}}... --name-only | sed 's|[^/]*$||' | sort -u >touched.txt &&
-          comm -12 directories.txt touched.txt >packages.txt &&
-          cat packages.txt &&
-          sed -e 's/\/$//' -e 's/.*/"&", /' -e '$s/, $//' <packages.txt >list.txt &&
-          echo "::set-output name=matrix::[$(tr -d '\n' <list.txt)]"
+          define_matrix matrix directories.txt touched.txt packages.txt &&
 
           git ls-files \*/release.sh | sed 's|[^/]*$||' | sort >releaseable.txt &&
-          comm -12 releaseable.txt touched.txt >artifacts.txt &&
-          cat artifacts.txt &&
-          sed -e 's/\/$//' -e 's/.*/"&", /' -e '$s/, $//' <artifacts.txt >list.txt &&
-          echo "::set-output name=artifacts::[$(tr -d '\n' <list.txt)]"
+          define_matrix artifacts releaseable.txt touched.txt artifacts.txt
     outputs:
       matrix: ${{ steps.set-matrix.outputs.matrix }}
       artifacts: ${{ steps.set-matrix.outputs.artifacts }}
   build-packages:
     needs: determine-packages
     runs-on: windows-latest
+    if: needs.determine-packages.outputs.matrix != '["empty"]'
```

Comment:
Do we need a string in our array to explicitly mark it as empty or could we just check for an empty array (`[]`)?

Reply:
If we use the empty array, then the matrix jobs will _still_ fail to even execute. I changed it to the empty string, and clarified in the commit message:

> Note: we need to add a bogus string because otherwise the parser that wants to construct the matrix jobs will _still_ fail.

Reply:
See here for the error: https://github.com/git-for-windows/build-extra/actions/runs/1780285695, and note that it uses the empty array in the workflow: https://github.com/git-for-windows/build-extra/actions/runs/1780285695/workflow

Reply:
Since the logs will be deleted in a couple months, here is a [link to the `main.yml` revision](https://github.com/git-for-windows/build-extra/blob/db2e236e66dbe84c9ba0faf66075119faf3fe4e2/.github/workflows/main.yml) and a screenshot for posterity:

![image](https://user-images.githubusercontent.com/127790/152189287-1998fd36-1071-4633-85fe-1d158d3aa8d0.png)

Reply:
@rimrul does that address your concern?

Reply:
Yes.